### ISSUE TITLE
AddressSanitizer: NULL-pointer READ at XMPMeta.cpp:1134

### ISSUE THREAD

**Issue by HongxuChen** [Exiv2/exiv2#984]
**Describe the bug**
`exiv2json` may crash due to NULL-pointer read.

**To Reproduce**
Steps to reproduce the behaviour:
1. run `exiv2json $f` where `exiv2json` is compiled with AddressSanitizer.
2. the program crashes with the backtrace like:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==25518==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x00000042a51b bp 0x7fffffffaae0 sp 0x7fffffffaa80 T0)
==25518==The signal is caused by a READ memory access.
==25518==Hint: address points to the zero page.
    #0 0x42a51a in __asan::asan_delete(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) (/home/hongxu/FOT/exiv2-all/exiv2-asan/BUILD/bin/exiv2json+0x42a51a)
    #1 0x517a72 in operator delete(void*) (/home/hongxu/FOT/exiv2-all/exiv2-asan/BUILD/bin/exiv2json+0x517a72)
    #2 0x7ffff73b6232 in std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::~pair() /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_pair.h:193:56
    #3 0x7ffff73b6208 in void __gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::destroy<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/ext/new_allocator.h:140:28
    #4 0x7ffff73b61a7 in void std::allocator_traits<std::allocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::destroy<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(std::allocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/alloc_traits.h:487:8
    #5 0x7ffff73b610b in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_destroy_node(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_tree.h:650:2
    #6 0x7ffff73b60ae in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_drop_node(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_tree.h:658:2
    #7 0x7ffff75d4377 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_erase_aux(std::_Rb_tree_const_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_tree.h:2477:7
    #8 0x7ffff77eaafa in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::erase[abi:cxx11](std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_tree.h:1125:2
    #9 0x7ffff77d0270 in std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::erase[abi:cxx11](std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_map.h:1031:21
    #10 0x7ffff77c95bf in XMPMeta::DeleteNamespace(char const*) /home/hongxu/FOT/exiv2-all/exiv2-asan/xmpsdk/src/XMPMeta.cpp:1134:28
    #11 0x7ffff7753c82 in WXMPMeta_DeleteNamespace_1 /home/hongxu/FOT/exiv2-all/exiv2-asan/xmpsdk/src/WXMPMeta.cpp:285:3
    #12 0x7ffff75bac98 in TXMPMeta<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::DeleteNamespace(char const*) /home/hongxu/FOT/exiv2-all/exiv2-asan/xmpsdk/include/client-glue/TXMPMeta.incl_cpp:280:2
    #13 0x7ffff75af496 in Exiv2::XmpParser::registerNs(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/hongxu/FOT/exiv2-all/exiv2-asan/src/xmp.cpp:559:13
    #14 0x7ffff750ebb5 in Exiv2::XmpProperties::registeredNamespaces(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&) /home/hongxu/FOT/exiv2-all/exiv2-asan/src/properties.cpp:2663:14
    #15 0x520c02 in main /home/hongxu/FOT/exiv2-all/exiv2-asan/samples/exiv2json.cpp:328:17
    #16 0x7ffff5b71b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #17 0x41e8b9 in _start (/home/hongxu/FOT/exiv2-all/exiv2-asan/BUILD/bin/exiv2json+0x41e8b9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/home/hongxu/FOT/exiv2-all/exiv2-asan/BUILD/bin/exiv2json+0x42a51a) in __asan::asan_delete(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType)
==25518==ABORTING
```

Other POCs and error messages are available [here](https://github.com/ntu-sec/pocs/tree/master/exiv2/1b2e2f40/crashes).

**Expected behavior**
Exit with non-zero.

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04.2 LTS
 - Compiler & Version: Clang-6
 - Compilation mode and/or compiler flags: "-fsanitize=address"

**Additional context**
- POC files are corrupted "Progressive Graphics image data".
```
$ `ldd ./exiv2json`
        linux-vdso.so.1 (0x00007ffff7ffa000)
        libexiv2.so.27 => /home/hongxu/FOT/exiv2-all/exiv2-asan/BUILD/lib/libexiv2.so.27 (0x00007ffff70c8000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007ffff6eab000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ffff6b22000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffff6784000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ffff6565000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ffff635d000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ffff6159000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ffff5f41000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffff5b50000)
        libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007ffff591e000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ffff7dd5000)
```

**Comment by clanmills** [comment]
Thank You for reporting this issue.  It have reproduced this and it will be investigated in the next few days.  We have a "dot" release v0.27.3 scheduled for 2019-09-30 and I hope it will include a fix for this issue.

**Comment by clanmills** [comment]
I haven't finishing figuring this out.  For sure, there's a "quick fix" by removing the call to `Exiv2::XmpProperties::registeredNamespaces(nsDict)`

However that does solve the underlying matter which has something to do will registering namespaces.

```
$ git diff
diff --git a/samples/exiv2json.cpp b/samples/exiv2json.cpp
index a33a383b..9b8d1a0c 100644
--- a/samples/exiv2json.cpp
+++ b/samples/exiv2json.cpp
@@ -325,7 +325,7 @@ int main(int argc, char* const argv[])
 
                 // get the namespace dictionary from XMP
                 Exiv2::Dictionary                          nsDict;
-                Exiv2::XmpProperties::registeredNamespaces(nsDict);
+                // Exiv2::XmpProperties::registeredNamespaces(nsDict);
 
                 // create and populate a Jzon::Object for the namespaces
                 Jzon::Object    xmlns;
$ bin/exiv2json x ~/xmpCrash.pgf 
{
	"Xmp": {
		"tiff": {
			"Software": "digiKam-1.0.0-beta2 (rev. : 979154)",
			"ImageWidth": "320",
			"ImageLength": "211",
			"PixelXDimension": "320",
			"PixelYDimension": "211"
		},
		"xmp": {
			"CreatorTool": "digiKam-1.0.0-beta2 (rev. : 979154)"
		},
		"xmlns": {
			"tiff": "",
			"xmp": ""
		}
	}
}
$ 
```

I can extract the XMP from the file as follows:
```
$ dmpf ~/xmpCrash.pgf | grep -A 2 -B 2 begin
  0x63e0    25568: ..iTXtXML:com.ad  ->  0a 8d 69 54 58 74 58 4d 4c 3a 63 6f 6d 2e 61 64
  0x63f0    25584: obe.xmp.....<?xp  ->  6f 62 65 2e 78 6d 70 00 00 00 00 00 3c 3f 78 70
  0x6400    25600: acket begin="...  ->  61 63 6b 65 74 20 62 65 67 69 6e 3d 22 ef bb bf
  0x6410    25616: " id="W5M0MpCehi  ->  22 20 69 64 3d 22 57 35 4d 30 4d 70 43 65 68 69
  0x6420    25632: HzreSzNTczkc9d"?  ->  48 7a 72 65 53 7a 4e 54 63 7a 6b 63 39 64 22 3f
$ dmpf ~/xmpCrash.pgf | grep -A 2 -B 2 end
  0x6e40    28224:    .              ->  20 20 20 0a 20 20 20 20 20 20 20 20 20 20 20 20
  0x6e50    28240:                .  ->  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 0a
  0x6e60    28256: <?xpacket end="w  ->  3c 3f 78 70 61 63 6b 65 74 20 65 6e 64 3d 22 77
  0x6e70    28272: "?>=.G+....sRGB.  ->  22 3f 3e 3d 89 47 2b 00 00 00 01 73 52 47 42 00
  0x6e80    28288: ........pHYs....  ->  ae ce 1c e9 00 00 00 09 70 48 59 73 00 00 0b 13
$ 
```
I can extract and format it with:
```
$ dd bs=1 skip=$((25600-4)) count=$((28272+3-(25600-4))) if=~/xmpCrash.pgf 2>/dev/null | xmllint --pretty 2 -
<?xml version="1.0"?>
<?xpacket
begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta
    xmlns:x="adobe:ns:meta/"
    x:xmptk="XMP Core 4.1.1-Exiv2"
  >
 <rdf:RDF
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
  <rdf:Description
        xmlns:tiff="http://ns.adobe.com/tiff/1.0/"
        xmlns:xap="http://ns.adobe.com/xap/1.0/"
        xmlns:exif="http://ns.adobe.com/tiff/1.0/"
        rdf:about=""
        tiff:Software="digiKam-1.0.0-beta2 (rev.&#xA0;: 979154)"
        tiff:ImageWidth="320"
        tiff:ImageLength="211"
        xap:CreatorTool="digiKam-1.0.0-beta2 (rev.&#xA0;: 979154)"
        exif:PixelXDimension="320"
        exif:PixelYDimension="211"
    />
 </rdf:RDF>
</x:xmpmeta>
<?xpacket
end="w"?>
$ dd bs=1 skip=$((25600-4)) count=$((28272+3-(25600-4))) if=~/xmpCrash.pgf | exiv2 -px -
Xmp.tiff.Software                            XmpText    36  digiKam-1.0.0-beta2 (rev. : 979154)
Xmp.tiff.ImageWidth                          XmpText     3  320
Xmp.tiff.ImageLength                         XmpText     3  211
Xmp.tiff.PixelXDimension                     XmpText     3  320
Xmp.tiff.PixelYDimension                     XmpText     3  211
Xmp.xmp.CreatorTool                          XmpText    36  digiKam-1.0.0-beta2 (rev. : 979154)
1140 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```
I think this is being caused by the namespace `xap` being defined in the file as `http://ns.adobe.com/xap/1.0/`

However Exiv2 knows `http://ns.adobe.com/xap/1.0/` and namespace `xmp`
```
$ exiv2 -vV | grep http://ns.adobe.com/xap/1.0/$
xmlns=xmp:http://ns.adobe.com/xap/1.0/
$ 
```
For sure we shouldn't be crashing.  I think we're releasing memory that's not in the heap.

**Comment by clanmills** [comment]
This is being caused by:
```
        xmlns:tiff="http://ns.adobe.com/tiff/1.0/"
        xmlns:xap="http://ns.adobe.com/xap/1.0/"
        xmlns:exif="http://ns.adobe.com/tiff/1.0/"
```
Both `exif` and `tiff` refer to the same namespace which is `http://ns.adobe.com/tiff/1.0/`

I've modified npe_read_XMPMeta.cpp-1134_1.txt, so that, it would be `xmlns:tiff="http://ns.adobe.com/tiff/1.0/` and `xmlns:exif="http://ns.adobe.com/exif/1.0/` and the crash goes away.

```
539 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/asan_build $ diff ~/temp/npe*
1622c1622
<   0x6550    25936: dobe.com/tiff/1.  ->  64 6f 62 65 2e 63 6f 6d 2f 74 69 66 66 2f 31 2e
---
>   0x6550    25936: dobe.com/exif/1.  ->  64 6f 62 65 2e 63 6f 6d 2f 65 78 69 66 2f 31 2e
540 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/asan_build $ 
```
The predefined values are:
```
547 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/asan_build $ bin/exiv2 -vV | grep adobe | grep exif
xmlns=aux:http://ns.adobe.com/exif/1.0/aux/
xmlns=exif:http://ns.adobe.com/exif/1.0/
548 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/asan_build $ bin/exiv2 -vV | grep adobe | grep tiff
xmlns=tiff:http://ns.adobe.com/tiff/1.0/
549 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/asan_build $ 
```

Curiously the code in xmpsdk has a warning about "Don't allow standard namespaces to be deleted".
```
// *** Don't allow standard namespaces to be deleted.
// *** We would be better off not having this. Instead, have local namespaces from parsing be
// *** restricted to the object that introduced them.

/* class-static */ void
XMPMeta::DeleteNamespace ( XMP_StringPtr namespaceURI )
{
	XMP_StringMapPos uriPos = sNamespaceURIToPrefixMap->find ( namespaceURI );
	if ( uriPos == sNamespaceURIToPrefixMap->end() ) return;

	XMP_StringMapPos prefixPos = sNamespacePrefixToURIMap->find ( uriPos->second );
	XMP_Assert ( prefixPos != sNamespacePrefixToURIMap->end() );
	sNamespaceURIToPrefixMap->erase ( uriPos );
	sNamespacePrefixToURIMap->erase ( prefixPos );

}	// DeleteNamespace
```

XMPsdk maintains two maps `URIToPrefixMap` to `PrefixToURIMap` and it's crashing on `sNamespacePrefixToURIMap->erase ( prefixPos )`

More work required to understand both this code and what's wrong.

**Comment by clanmills** [comment]
It's the double mapping.  The mapping prefix <-> URI and URI <-> prefix must be unique.  

I have a fix.  I would like to throw an XMPError, however that causes chaos in the test suite.  
```
/* class-static */ void
XMPMeta::RegisterNamespace ( XMP_StringPtr   namespaceURI,
                             XMP_StringPtr   prefix )
{
    if ( (*namespaceURI == 0) || (*prefix == 0) ) {
        XMP_Throw ( "Empty namespace URI or prefix", kXMPErr_BadParam );
    }

    XMP_VarString   nsURI ( namespaceURI );
    XMP_VarString   prfix ( prefix );
    if ( prfix[prfix.size()-1] != ':' ) prfix += ':';
    VerifySimpleXMLName ( prefix, prefix+prfix.size()-1 );  // Exclude the colon.

    int  nRegistered   = 0 ;
    bool bThrowEnabled = false; // don't enable throw - it causes trouble in the test suite

    // register namespace if not known
    if( sNamespaceURIToPrefixMap->find(nsURI) == sNamespaceURIToPrefixMap->end()
    &&  sNamespacePrefixToURIMap->find(prfix) == sNamespacePrefixToURIMap->end()
    ) {
        // Set the new namespace in both maps.
        (*sNamespaceURIToPrefixMap)[nsURI] = prfix;
        (*sNamespacePrefixToURIMap)[prfix] = nsURI;
        nRegistered = 2 ;
    } else {
        if (sNamespaceURIToPrefixMap->find(nsURI) != sNamespaceURIToPrefixMap->end()) {
            if ( (*sNamespaceURIToPrefixMap)[nsURI] == prfix ) {
                nRegistered ++ ;
            }
        }
        if (sNamespaceURIToPrefixMap->find(nsURI) != sNamespaceURIToPrefixMap->end()) {
            if ( (*sNamespacePrefixToURIMap)[prfix] == nsURI ) {
                nRegistered ++ ;
            }
        }
    }

    if ( nRegistered != 2 && bThrowEnabled ) {
        XMP_Throw ( "Illegal namespace/prefix", kXMPErr_BadParam );
    }

}   // RegisterNamespace
```

I've inspected the code for XMPMeta::RegisterNamespace in XMP-Toolkit-SDK-CC201607 and it's quite different.  We can't simply copy the 2016 code for a fix.

If we disable throwing an error, the following happens:

1) We (correctly) throw an XMPerror concerning namespace map problem:
```
.../0.27-maintenance/build $ bin/exiv2json  ~/Downloads/npe_read_XMPMeta.cpp-1134_1.orig.pgf 
Caught Exiv2 exception 'XMP Toolkit error 9: Fatal namespace map problem'
.../0.27-maintenance/build $ 
```

2) We get the following errors from newtests:
```
.../0.27-maintenance/build $ (cd ../test ; make newtests 2>&1 | grep ^FAIL:)
FAIL: test_run (bugfixes.github.test_CVE_2017_14857.TestCvePoC)
FAIL: test_run (bugfixes.github.test_issue_851.DenialOfServiceInAdjustTimeOverflow)
FAIL: test_run (bugfixes.redmine.test_issue_751.AdobeXmpNamespace)
FAIL: test_run (bugfixes.redmine.test_issue_937.DarwinCoreXmpMetadataPrint)
.../0.27-maintenance/build $ 
```

3) xmpparser-test.sh is producing errors.

I will investigate the errors in new tests and xmpparser-test.sh

**Comment by clanmills** [comment]
I know why xmpparse-test.sh is failing.  My "fix" above guarantees that both prefix and url and unique.  You cannot have more than one prefix referring to a url.  However the file `test/ data/StaffPhotographer-Example.xmp` contradicts the exif built-in value for `http://ns.adobe.com/xap/1.0/mm/`.  The file uses the prefix `xapMM` and the built-in prefix is `xmpMM`

```
529 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ xmllint --pretty 2 data/StaffPhotographer-Example.xmp | grep xmlns | grep MM
        xmlns:xapMM="http://ns.adobe.com/xap/1.0/mm/"
530 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ ../build/bin/exiv2 -vVg xmlns | grep MM
xmlns=xmpMM:http://ns.adobe.com/xap/1.0/mm/
531 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ 
```
This issue applies to several namespaces:
```
539 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ xmllint --pretty 2 data/StaffPhotographer-Example.xmp | grep xmlns:xap
        xmlns:xap="http://ns.adobe.com/xap/1.0/"
        xmlns:xapMM="http://ns.adobe.com/xap/1.0/mm/"
        xmlns:xapRights="http://ns.adobe.com/xap/1.0/rights/"
540 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ 
```

Adobe use the prefix xmpXX, never xapXX
```
536 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ ../build/bin/exiv2 -vVg xmlns | grep =xmp
xmlns=xmp:http://ns.adobe.com/xap/1.0/
xmlns=xmpBJ:http://ns.adobe.com/xap/1.0/bj/
xmlns=xmpDM:http://ns.adobe.com/xmp/1.0/DynamicMedia/
xmlns=xmpG:http://ns.adobe.com/xap/1.0/g/
xmlns=xmpGImg:http://ns.adobe.com/xap/1.0/g/img/
xmlns=xmpMM:http://ns.adobe.com/xap/1.0/mm/
xmlns=xmpNote:http://ns.adobe.com/xmp/note/
xmlns=xmpRights:http://ns.adobe.com/xap/1.0/rights/
xmlns=xmpT:http://ns.adobe.com/xap/1.0/t/
xmlns=xmpTPg:http://ns.adobe.com/xap/1.0/t/pg/
xmlns=xmpidq:http://ns.adobe.com/xmp/Identifier/qual/1.0/
537 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ 
```

The fix for this is to modify the prefix definitions in `StaffPhotographer-Example.xmp`, rerun the test and update the reference output in `test/data/data/xmpparser-test.out` 

This is a little alarming!  The whole point of a namespace prefix is an alias within a file and it should not have global consequences.  We should have two namespace maps.  A global map and a file instance map.  This is a matter for the Adobe XMPsdk and Exiv2 should not have to consider this.

The issue really is "why does exiv2json crash?".  I think it's attempting to deallocate memory that's not in the heap.  Somehow there are prefix and/or url strings in the maps that reference non-heap memory.  How can that happen?

I haven't investigate the failures in 'newtests'.

Lots more work ahead, so unlikely to be fixed in Exiv2 v0.27.3

**Comment by clanmills** [comment]
I know why it's crashing.  It's attempting to return static memory to the heap.  It's in this definition in `include/exiv2/properties.hpp` _line 68_:
```
    struct EXIV2API XmpNsInfo {
        //! For comparison with prefix
        struct Prefix {
            //! Constructor.
            explicit Prefix(const std::string& prefix);
            //! The prefix string.
            std::string prefix_;
        };
        //! For comparison with namespace
        struct Ns {
            //! Constructor.
            explicit Ns(const std::string& ns);
            //! The namespace string
            std::string ns_;
        };
        //! Comparison operator for namespace
        bool operator==(const Ns& ns) const;
        //! Comparison operator for prefix
        bool operator==(const Prefix& prefix) const;

        const char* ns_;                //!< Namespace
        const char* prefix_;            //!< (Preferred) prefix
        const XmpPropertyInfo* xmpPropertyInfo_; //!< List of known properties
        const char* desc_;              //!< Brief description of the namespace
    };
```
Being used from in `src/properties.cpp` _line105_:
```
    extern const XmpNsInfo xmpNsInfo[] = {
        // Schemas   -   NOTE: Schemas which the XMP-SDK doesn't know must be registered in XmpParser::initialize - Todo: Automate this
        { "http://purl.org/dc/elements/1.1/",             "dc",             xmpDcInfo,        N_("Dublin Core schema")                        },
        { "http://www.digikam.org/ns/1.0/",               "digiKam",        xmpDigikamInfo,   N_("digiKam Photo Management schema")           },
        { "http://www.digikam.org/ns/kipi/1.0/",          "kipi",           xmpKipiInfo,      N_("KDE Image Program Interface schema")        },
        { "http://ns.adobe.com/xap/1.0/",                 "xmp",            xmpXmpInfo,       N_("XMP Basic schema")                          },
        { "http://ns.adobe.com/xap/1.0/rights/",          "xmpRights",      xmpXmpRightsInfo, N_("XMP Rights Management schema")              },
        { "http://ns.adobe.com/xap/1.0/mm/",              "xmpMM",          xmpXmpMMInfo,     
...
```
The fix for this is to use a constructor for `struct XmpNsInfo` to ensure that ns_ and other data members of the structure are allocated on the heap. **BUT** this data structure is in the API and shouldn't be modified!  So, I'll investigate duplicating this data structure into heap memory before running the loop to register those namespaces.

In `src/properties.cpp` _line 2663_, we __could__ add an `strdup()` to the ns, prefix arguments.  However that will leak as there's nothing to clean it up and I can't do that in ~XmpNsInfo:: XmpNsInfo without changing the API.  Yuck.

```
    void XmpProperties::registeredNamespaces(Exiv2::Dictionary& nsDict)
    {
        for (unsigned int i = 0; i < EXV_COUNTOF(xmpNsInfo); ++i) {
             Exiv2::XmpParser::registerNs(xmpNsInfo[i].ns_,xmpNsInfo[i].prefix_);
        }
        Exiv2::XmpParser::registeredNamespaces(nsDict);
    }
```

I'm looking in the wrong place.  It's XMPsdk who is doing the delete.  It's wrong to require `XMPMeta::RegisterNamespace` to be called with addresses on the heap.  XMPsdk must build the maps in heap memory.

**Comment by clanmills** [comment]
Fixed in 'master' with #1093 Fixed in '0.27-maintenance' with #1125
